pennylane/devices/null_qubit.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+The null.qubit device is a no-op device for benchmarking PennyLane's auxiliary functionality outside direct circuit evaluations.
+"""
+from collections import defaultdict
+
+from pennylane.ops.qubit.attributes import diagonal_in_z_basis
+
+from pennylane.devices import DefaultQubit
+from .._version import __version__
+
+# pylint: disable=unused-argument
+class NullQubit(DefaultQubit):
+    """Null qubit device for PennyLane.
+
+    Args:
+        wires (int, Iterable[Number, str]): Number of subsystems represented by the device,
+            or iterable that contains unique labels for the subsystems as numbers (i.e., ``[-1, 0, 2]``)
+            or strings (``['auxiliary', 'q1', 'q2']``). Default 1 if not specified.
+    """
+
+    name = "Null qubit PennyLane plugin"
+    short_name = "null.qubit"
+    pennylane_requires = __version__
+    version = __version__
+    author = "Xanadu Inc."
+
+    def __init__(self, wires, *args, **kwargs):
+        defaultKwargs = {"shots": None}
+        kwargs = {**defaultKwargs, **kwargs}
+
+        self._operation_calls = defaultdict(int)
+        self._shots = kwargs["shots"]
+        self._shot_vector = None
+        self.custom_expand_fn = None
+        super().__init__(wires=wires, shots=self._shots)
+
+    # pylint: disable=arguments-differ
+    def apply(self, operations, *args, **kwargs):
+        for op in operations:
+            self._apply_operation(None, op)
+
+    def _apply_operation(self, state, operation):
+        if operation.__class__.__name__ in self._apply_ops:
+            return self._apply_ops[operation.base_name](state, axes=None, inverse=operation.inverse)
```

Comment:
Not sure why we need to keep the `self._apply_ops` around. 

Why not just have this function do purely: `self._operation_calls[operation.name] += 1` ?

Reply:
At this point of development, we have each of the special methods in the _apply_ops list defined, and we perform the +1 to the _operation_calls list there. In this way, applying the gate either with `apply` or by directly calling the method will execute the +1 in the list. I saw it being used in both ways during development or debugging, and because of that, I decided to keep it like that. Alternatively, we can just `pass` the special methods and execute the +1 in `_apply_operation`, or maybe even directly in `apply` and also pass the `_apply_operation` method. @mlxd and @albi3ro what are your thought about that?

Reply:
I'm defining it to mimic the `default.qubit` as answered below.
My initial intention was to pass the special methods and execute the +1 in _apply_operation as later suggested by Christina, but I had the impression that keeping as I found could be a better idea.
Anyways, I will improve this part.